core/src/main/scala/org/apache/spark/ui/viz/VisualizationListener.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ui.viz
+
+import scala.collection.mutable
+import scala.xml.{Node, Unparsed}
+
+import org.apache.spark.SparkConf
+import org.apache.spark.scheduler._
+import org.apache.spark.ui.SparkUI
+
+/**
+ * A SparkListener that constructs RDD DAG visualization for the UI.
+ */
+private[ui] class VisualizationListener(conf: SparkConf) extends SparkListener {
```

Comment:
After some more thought, I think that currently this is pretty tied to Job/Dag visualization, so it could make sense to just leave it as-is. If we want to expose something more general later, we can do that. Maybe add a NOTE that says if we want to make it public later on we should consider giving a more general representation of the scoping concept.